src/NativeScript/inlineFunctions.js

```diff
@@ -14,6 +14,103 @@ Object.assign(global, {
     NSMakeRange(location, length) {
         return new NSRange({ location, length });
     },
+
+    // Decorators support    
+    __decorate(decorators, target, key, desc) {
+        var c = arguments.length;
+        var r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc;
+        var d;
+
+        if (typeof global.Reflect === "object" && typeof global.Reflect.decorate === "function") {
+            r = global.Reflect.decorate(decorators, target, key, desc);
+        }
+        else {
+            for (var i = decorators.length - 1; i >= 0; i--) {
+                if (d = decorators[i]) {
+                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
+                }
+            }
+        }
+        return c > 3 && r && Object.defineProperty(target, key, r), r;
+    },
+    __param(paramIndex, decorator) {
+        return function (target, key) {
+            decorator(target, key, paramIndex);
+        }
+    },  
+
+    // Decorators  
+    ObjC(name) {
+        var additionalParams = [];
+        var hasName = (name !== undefined && typeof name === "string");
+        var argsShift = hasName ? 1 : 0;
+        for (var i = argsShift; i < arguments.length; i++) {
+            additionalParams[i - argsShift] = arguments[i];
+        }
+
+        return function (target, propertyKey, descriptor) {
+            if (propertyKey === undefined) { // Class decorator
+                var implementation = {};
+                var baseTarget = Object.getPrototypeOf(target);
+                
+                for (var prop in target.prototype) {
+                    if (target.prototype.hasOwnProperty(prop)) {
+                        implementation[prop] = target.prototype[prop];
+                    }
+                }
+
+                var newName = hasName ? name : target.name;                
+                var extended = baseTarget.extend(implementation, {
```

Comment:
The benefit of using the `baseTarget.extend` is that we can specify the `name` of the class, which with the TS extends method is impossible. Also it will allow the TS extending to match and use the JS extend, so if any changes are made to the JS extend they can easily flow over to TS. Now the problem is that because of the TS extends, this creates two instances of the class. 

You can take a look at the suggestion by @PanayotCankov here: https://github.com/NativeScript/NativeScript/issues/4018#issuecomment-295600104 . Having the aggregated `@ObjC` decorator looked like a good idea initially, but last night when I tried to make the definitions, it was impossible to make them strongly typed. So I think now having different methods is the better option. But IMO the `ObjCMethod` one should just accept the things concerning the method return type/name, and not accept a JSON with the parameters and everything. Also this creates some confusion to the end user. They will need to know if the param decorator or the method signature takes precedence.